extensions/interactions/ImageClickInput/directives/oppia-interactive-image-click-input.component.ts

```diff
@@ -0,0 +1,284 @@
+      .getStaticImageUrl(AppConstants.LOADING_INDICATOR_URL);
+    this.isLoadingIndicatorShown = false;
+    this.isTryAgainShown = false;
+    this.dimensions = (
+      this.imagePreloaderService.getDimensionsOfImage(this.filepath));
+    this.imageContainerStyle = {
+      height: this.dimensions.height + 'px',
+      width: this.dimensions.width + 'px'
+    };
+    if (this.imagePreloaderService.inExplorationPlayer()) {
+      this.isLoadingIndicatorShown = true;
+      // For aligning the gif to the center of it's container.
+      const loadingIndicatorSize = (
+                (this.dimensions.height < 124) ? 24 : 120);
+      this.imageContainerStyle = {
+        height: this.dimensions.height + 'px'
+      };
+      this.loadingIndicatorStyle = {
+        height: loadingIndicatorSize + 'px',
+        width: loadingIndicatorSize + 'px'
+      };
+      this.loadImage();
+    } else {
+      // This is the case when user is in exploration editor or in
+      // preview mode. We don't have loading indicator or try again for
+      // showing images in the exploration editor or in preview mode. So
+      // we directly assign the url to the imageUrl.
+      this.imageUrl = this.assetsBackendApiService.getImageUrlForPreview(
+        this.contextService.getEntityType(), this.contextService.getEntityId(),
+        this.filepath);
+    }
+
+    this.mouseX = 0;
+    this.mouseY = 0;
+    this.interactionIsActive = (this.lastAnswer === null);
+
+    this.currentlyHoveredRegions = [];
+    this.allRegions = this.imageAndRegions.labeledRegions;
+    if (!this.interactionIsActive) {
+      /**
+       * The following lines highlight the learner's last answer for
+       * this card. This need only be done at the beginning as if he
+       * submits an answer, based on newCardAvailable, the image
+       * is made inactive, so his last selection would be highlighted.
+       */
+      this.mouseX = this.lastAnswer.clickPosition[0];
+      this.mouseY = this.lastAnswer.clickPosition[1];
+      this.updateCurrentlyHoveredRegions();
+    }
+
+    this.currentInteractionService.registerCurrentInteraction(null, null);
+  }
+
+  loadImage(): void {
+    this.imagePreloaderService.getImageUrl(this.filepath)
+      .then((objectUrl: string) => {
+        this.isTryAgainShown = false;
+        this.isLoadingIndicatorShown = false;
+        this.imageUrl = objectUrl;
+      }, () => {
+        this.isTryAgainShown = true;
+        this.isLoadingIndicatorShown = false;
+      });
+  }
+
+  updateCurrentlyHoveredRegions(): void {
+    for (let i = 0; i < this.imageAndRegions.labeledRegions.length; i++) {
+      const labeledRegion = this.imageAndRegions.labeledRegions[i];
+      const regionArea = labeledRegion.region.area;
+      if (this.mouseX >= regionArea[0][0] &&
+          this.mouseX <= regionArea[1][0] &&
+          this.mouseY >= regionArea[0][1] &&
+          this.mouseY <= regionArea[1][1]) {
```

Comment:
I still think it wouldn't hurt to have this in a separate function (something like `isMouseInsideRegion(regionArea)`), the readability of the code improves, also this function is testable separately.

Reply:
Done!